docs/architecture/psa-thread-safety.md

```diff
@@ -67,16 +67,32 @@ In the medium to long term, performing a slow or blocking operation (for example
 
 We may want to go directly to a more sophisticated approach because when a system works with a global lock, it's typically hard to get rid of it to get more fine-grained concurrency.
 
+### Key destruction short-term requirements
+
+#### Summary of guarantees in the short term
+
+When `psa_destroy_key` returns:
+
+1. The key identifier doesn't exist. Rationale: this is a functional requirement for persistent keys: the caller can immediately create a new key with the same identifier.
+2. The resources from the key have been freed. Rationale: in a low-resource condition, this may be necessary for the caller to re-create a similar key, which should be possible.
+3. The call must not block indefinitely, and in particular cannot wait for an event that is triggered by application code such as calling an abort function. Rationale: this may not strictly be a functional requirement, but it is an expectation `psa_destroy_key` does not block forever due to another thread, which could potentially be another process on a multi-process system. (This doesn't mean that the call can't block at all. It is acceptable for example to block waiting for another thread to complete an API call that it had already started, but it is unacceptable to wait for another thread to make an API call.)
```

Comment:
Could we go so far as saying that the call must not block but to wait for another thread or process to complete a PSA cryptography API call?

Reply:
Even in the absence of drivers, I don't know how we would implement that with only mutexes and condition variables. We would need signals and signal handlers as well. Am I missing something? Do we want to introduce in our threading abstraction layer those as well? (Doesn't sound like something that is easy to do in a portable way.)

Reply:
I think the only realistic solution would be to copy the key material so that drivers don't get a pointer to it.

Reply:
So `destroy_key` would erase (but not free) the copy directly and the operation thread would calculate incorrect results but not do any harm. It would also block until `destroy_key` finishes (to keep the memory being erased valid), at which point it can invalidate the operation and exit gracefully). Right?

Reply:
This would work with static key material, but for example with HMAC the operation thread writes the buffer in question as well. It is unlikely, but conceivable that the instructions of the two write operation will be scheduled in a way that the key material (or parts of it) remain in the buffer after `destroy_key` erases it.

Reply:
> Even in the absence of drivers, I don't know how we would implement that with only mutexes and condition variables. We would need signals and signal handlers as well. Am I missing something? Do we want to introduce in our threading abstraction layer those as well? (Doesn't sound like something that is easy to do in a portable way.)

I am a bit lost here. In the current version we have "It is acceptable for example to block waiting for another thread to complete an API call that it had already started, but it is unacceptable to wait for another thread to make an API call." thus by proposing "Could we go so far as saying that the call must not block but to wait for another thread or process to complete a PSA cryptography API call?" I was not proposing anything new, I was just wondering if we could restrict the cases where we can block in the implementation of `psa_destroy_key()` to that one case.
Ah maybe it id because I said "for another thread OR PROCESS"? Then just ignore "OR PROCESS", my bad.

Reply:
I am sorry, it is my bad, my brain skipped the word "but" and I read "Could we go so far as saying that the call must not block to wait for another thread or process to complete a PSA cryptography API call?" and I was reacting to that.